File: src/main/java/com/github/iamniklas/liocore/led/LEDStripManager.java
package com.github.iamniklas.liocore.led;

import com.github.iamniklas.liocore.config.ProgramConfiguration;
import com.github.iamniklas.liocore.led.colorspace.LIOColor;
import com.github.iamniklas.liocore.procedures.*;

public class LEDStripManager implements ProcedureCalls {
    public static int ledCount;
    public final LEDStrip ledStrip;

    public ProcContainer procContainer = new ProcContainer(this);

    private final LEDRenderer renderer;

    public LEDStripManager(LEDRenderer _renderer, int _ledCount) {
        renderer = _renderer;
        ledCount = _ledCount;

        ledStrip = new LEDStrip(ledCount);
    }

    public void update() {
        procContainer.update();
        procContainer.postUpdate();
    }

    public void render() {
        renderer.render(ledStrip.getStripData());
    }

    public void waitFrametime() {
        try {
            Thread.sleep(ProgramConfiguration.configuration.frametime);
        }
        catch (InterruptedException ie) {
            ie.printStackTrace();
        }
    }

    public void setPixel(int _index, LIOColor _color) {
        ledStrip.setPixel(_index, _color);
    }
    public void setPixel(int _index, int _r, int _g, int _b) {
        ledStrip.setPixel(_index, new LIOColor(_r, _g, _b));
    }

    public void setArea(int _start, int _end, LIOColor _color) {
        for (int i = _start; i < _end; i++) {
            ledStrip.setPixel(i, _color);
        }
    }

    public void setAllPixels(LIOColor _color) {
        for (int i = 0; i < ledCount; i++) {
            ledStrip.setPixel(i, _color);
        }
    }
    public void setAllPixels(int _r, int _g, int _b) {
        LIOColor color = new LIOColor(_r, _g, _b);
        for (int i = 0; i < ledCount; i++) {
            ledStrip.setPixel(i, color);
        }
    }

    @Override
    public void onProcedureStart(Procedure _procedure) {
        _procedure.start();
    }

    @Override
    public void onProcedureQueued() {}

    @Override
    public void onProcedureFinish() {}
}

File: src/main/java/com/github/iamniklas/liocore/led/LEDStripManager.kt
package com.github.iamniklas.liocore.led

import com.github.iamniklas.liocore.led.json.LEDStatus
import com.github.iamniklas.liocore.network.LEDChangeModel
import com.github.iamniklas.liocore.procedures.*
import java.awt.Color
import javax.swing.JLabel

class LEDStripManager(_leds: Array<JLabel?>, _clearOnExit: Boolean) : ProcedureCalls {
    private val mLedStripConfig = LEDStripConfig()

    var mStripData = LEDStrip(LED_COUNT)

    var mProcContainer = ProcContainer(this)
    var frametime = 16
    var mStrip = LEDStripRenderer()
    fun update() {
        mProcContainer.update()
        mProcContainer.postUpdate()
        for (i in 0 .. mStripData.mStrip.size) {
            mStrip.setPixel(i, mStripData.mStrip[i])
        }
        mStrip.render()
        try {
            Thread.sleep(frametime.toLong())
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    /**
     * @param _pxl The pixel index
     * @param _color The color to set the pixel
     */
    fun setPixel(_pxl: Int, _color: Color) {
        var _color = _color
        _color = Color(Math.max(0, _color.red), Math.max(0, _color.green), Math.max(0, _color.blue))
        mStripData.mStrip[_pxl] = _color
    }

    fun setPixel(_pxl: Int, _r: Int, _g: Int, _b: Int) {
        val c = Color(Math.max(0, _r), Math.max(0, _g), Math.max(0, _b))
        mStripData.mStrip[_pxl] = c
    }

    /**
     * @param _startPxl The start index for the area
     * @param _endPxl The end index for the area
     * @param _color The color to fill the strip area with
     */
    fun setArea(_startPxl: Int, _endPxl: Int, _color: Color?) {
        for (i in _startPxl .. _endPxl) {
            mStripData.mStrip[i] = _color
        }
    }

    /**
     * @param _color The color to fill the strip with
     */
    fun setAllPixels(_color: Color?) {
        for (i in 0 .. LED_COUNT) {
            mStripData.mStrip[i] = _color
        }
    }

    override fun onProcedureStart(_proc: Procedure?) {
        _proc!!.start()
    }

    override fun onProcedureQueued() {}
    override fun onProcedureFinish() {}

    companion object {
        const val LED_COUNT = 300
        var mLEDStatus = LEDStatus()
    }

    init {
        System.out.println("LED Strip \tINIT \tSTART")
        frametime = 16
        LEDStripRenderer.leds = _leds
        System.out.println("LED Strip \tINIT \tDONE")
        val changeModel = LEDChangeModel()
        val bundle = LEDDataBundle()
        bundle.ledStrip = this
        bundle.procedureCalls = this
        bundle.data =
            "indices = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 299]; function update(strip, procedure, step, totalSteps) {d = Math.abs(Math.sin(step * Math.PI/180)); for(i = 0; i < indices.length; i++) { strip.setPixel(indices[i], 0, parseInt(d * 255), 0); } if(step >= totalSteps) { procedure.finish(); } }"
        changeModel.mProcedure = ProcedureType.Javascript
        changeModel.mBundle = bundle
        val proc = ProcedureFactory.getProcedure(changeModel.mProcedure, changeModel.mBundle!!)!! as JavascriptProcedure
        mProcContainer.queueProcedure(proc)
    }
}